### PR TITLE
Add multi threading in server and broadcast to all client

### PR DESCRIPTION
**Pull request by bp7968h**
- Code refactor for server
- Add multiple client connection simultaneously
- Broadcast message to all clients

### CODE REVIEW

File: rusty_socket_server/src/connection.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+use std::net::TcpStream;
+use std::io::{Read, Write};
+use std::sync::{Arc, Mutex};
+use rusty_socket_core::{DataFrame, OpCode};
+
+pub struct Connection {}
+
+impl Connection {
+    pub fn handle_frames(mut stream: TcpStream, active_conn: Arc<Mutex<Vec<TcpStream>>>) {
```
Comment (Suranjan77) [inline rusty_socket_server/src/connection.rs:9]: Better to use hash map data structure for multicasting messages to group of clients. But this things should be deferred to the application implementing this protocol.

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,3 @@
 # Built artifacts
 /target
+/.idea
```
Comment (Suranjan77) [inline .gitignore:3]: Add this into .gitignore

File: rusty_socket_server/src/handshake/response_line.rs [hunk 3]
```diff
@@ -23,11 +23,11 @@ impl ResponseLine {
 
     pub fn build(request_key: &str) -> Self {
         let mut response_headers: HashMap<String, String> = HashMap::new();
-        let accpet_key = Self::generate_websocket_accept_key(request_key);
+        let accept_key = Self::generate_websocket_accept_key(request_key);
```
Comment (Suranjan77) [inline rusty_socket_server/src/handshake/response_line.rs:26]: I didn't know we can use Self to call methods without &self as method parameter. Nicely done.

File: rusty_socket_server/src/socket_server.rs [hunk 4]
```diff
@@ -1,26 +1,33 @@
...
-    connections: Vec<Connections>,
+    active_connections: Arc<Mutex<Vec<TcpStream>>>,
 }
 
 impl SocketServer {
-    pub fn new() -> Self {
-        SocketServer {
-            target: "127.0.0.1:8080".parse::<SocketAddrV4>().unwrap(),
-            connections: Vec::new(),
+    pub fn build(address: impl ToSocketAddrs) -> Result<Self> {
```
Comment (Suranjan77) [inline rusty_socket_server/src/socket_server.rs:15]: Usually server would be listening on localhost and specified port. Just a suggestion based on my assumption.

File: rusty_socket_server/src/socket_server.rs [hunk 5]
```diff
@@ -29,27 +36,39 @@ impl SocketServer {
...
         match handshake.request {
-            Some(_request) => {
+            Some(_) => {
                 stream
                     .write_all(handshake.response.to_string().as_bytes())
                     .expect("Failed to write success response");
                 stream.flush().expect("Failed to flush success stream");
 
-                self.handle_frames(stream);
+                match self.active_connections.try_lock() {
```
Comment (Suranjan77) [inline rusty_socket_server/src/socket_server.rs:59]: Again, I think protocol should be simple implementation. Complexities like this should be handled by implementing application.

File: rusty_socket_server/src/connection.rs (hunk 1 above)
Comment (bp7968h) [reply rusty_socket_server/src/connection.rs:9]: I have no idea haha, we can talk about it after you are back.

File: rusty_socket_server/src/handshake/response_line.rs (hunk 3 above)
Comment (bp7968h) [reply rusty_socket_server/src/handshake/response_line.rs:26]: So, they are called associated function, you call associated function with namespace to self, but function that contains &self as it's parameters are methods.

File: rusty_socket_server/src/socket_server.rs (hunk 4 above)
Comment (bp7968h) [reply rusty_socket_server/src/socket_server.rs:15]: Yes this will be changed

File: rusty_socket_server/src/socket_server.rs (hunk 5 above)
Comment (bp7968h) [reply rusty_socket_server/src/socket_server.rs:59]: Yes, I am bit confused cause we did not decide on how the application would implement, this is something we can talk about,

**Pull request merged**